src/bin/semver-crater.rs

```diff
@@ -0,0 +1,114 @@
+use clap::{Parser, ValueEnum};
+use crates_io_api::{CratesQuery, Sort, Version};
+use itertools::Itertools;
+use std::time::Duration;
+
+#[derive(ValueEnum, Clone, Debug)]
+enum ReleaseFilter {
+    CheckAllReleases,
+    SkipAdjacentPatches,
+    SkipAdjacentMinor,
+}
+
+impl ReleaseFilter {
+    fn filter(&self, versions: Vec<Version>) -> Vec<(Version, Version)> {
+        let versions = {
+            let mut filtered: Vec<Version> = vec![versions
+                .first()
+                .expect("received empty list of versions")
+                .clone()];
+            filtered.extend(versions.windows(3).filter_map(|vec| {
+                let mut iter = vec.iter();
+                let version_v0 = semver::Version::parse(&iter.next().unwrap().num).unwrap();
+                let v1 = iter.next().unwrap();
+                let version_v2 = semver::Version::parse(&iter.next().unwrap().num).unwrap();
+                assert!(iter.next().is_none());
+
+                let keep_middle = match self {
+                    Self::CheckAllReleases => true,
+                    Self::SkipAdjacentPatches => {
+                        version_v0.major != version_v2.major || version_v0.minor != version_v2.minor
+                    }
+                    Self::SkipAdjacentMinor => version_v0.major != version_v2.major,
+                };
+                match keep_middle {
+                    true => Some(v1.clone()),
+                    false => None,
+                }
+            }));
+            filtered.push(
+                versions
+                    .last()
+                    .expect("received empty list of versions")
+                    .clone(),
+            );
+            filtered
```

Comment:
I won't deny -- this code is ugly as hell. But it works. And I don't really have any ideas on how to make it nicer.